### PR TITLE
provider/amazon: handle VPC load balancers on Classic ASGs

### PR DESCRIPTION
**Pull request by anotherchrisberry**
Handles an edge case where an ASG in Classic is using a VPC load balancer, but that load balancer belongs to (based on its name) a different application.

When we cache the ASG, we look at its launch config to determine the load balancers (based on `loadBalancerNames` to derive its load balancer relationships, and generate load balancer entries based on the location of the ASG. If the load balancer is in VPC, we end up generating an invalid key, which does end up in search results (we'll show that there's a load balancer in both VPC and Classic with the same name); however, when we retrieve the application load balancers, we currently only try to look up the invalid (e.g. Classic) key, which drops out of the result set.

@cfieber PTAL

### CODE REVIEW

**Comment by anotherchrisberry** [comment]
resolves SPIN-1972

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy [hunk 1]
```diff
@@ -63,15 +63,19 @@ class AmazonLoadBalancerProvider implements LoadBalancerProvider<AmazonLoadBalan
...
-      keys.addAll(serverGroup.relationships[LOAD_BALANCERS.ns] ?: [])
+      Collection<String> serverGroupLoadBalancers = serverGroup.relationships[LOAD_BALANCERS.ns] ?: []
+      serverGroupLoadBalancers.each {
+          keys.add(it)
+          def vpcKey = it + ':vpc-'
+          keys.addAll(allLoadBalancerKeys.findAll { it.startsWith(vpcKey)})
+      }
+    }
+    def nameMatches = allLoadBalancerKeys.findAll {
+      it ==~ 'aws:.*:' + applicationName + '-.*' || it ==~ 'aws:.*:' + applicationName || it ==~ 'aws:.*:' + applicationName + ':.*'
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy]: is there a way to do this (or some of this) via filter identifiers on line 69?

Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy]: ah I see the lines this is replacing... never mind me

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy [hunk 2]
```diff
@@ -63,15 +63,19 @@ class AmazonLoadBalancerProvider implements LoadBalancerProvider<AmazonLoadBalan
     CacheData application = cacheView.get(APPLICATIONS.ns, Keys.getApplicationKey(applicationName))
 
     def applicationServerGroups = application ? resolveRelationshipData(application, SERVER_GROUPS.ns) : []
+    def allLoadBalancerKeys = cacheView.getIdentifiers(LOAD_BALANCERS.ns)
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy]: might still be worth doing `cacheView.filterIdentifiers(LOAD_BALANCERS.ns, Keys.getLoadBalancerKey(applicationName + '*', '*', '*', null)` which will map to a glob of `aws:loadBalancers:*:*:applicationName*`

**Comment by cfieber** [comment]
LGTM

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy (hunk 2 above)
Comment (anotherchrisberry) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonLoadBalancerProvider.groovy]: Only downside to that is it'll miss questionably named load balancers, i.e. ASG `fooproxy-v001` uses ELB `foo-frontend`

**Pull request merged**